File: Curse/ServerProject/src/main/java/Server/CurseServerImpl.java
package Server;

import Domain.Angajat;
import Domain.Cursa;
import Domain.Inscriere;
import Domain.Pilot;

import Repository.RepoAngajat;
import Repository.RepoCursa;
import Repository.RepoInscriere;
import Repository.RepoPilot;
import Services.ICursaObserver;

import Services.ICursaService;
import Services.ProjectException;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class CurseServerImpl implements ICursaService {
    private RepoAngajat iRepositoryEmployee;
    private RepoCursa iRepositoryCursa;
    private RepoInscriere iRepositoryInscriere;
    private RepoPilot iRepositoryPilot;
    private Map<String, ICursaObserver> loggedEmployees;
    private final int defaultThreadsNo = 5;

    public CurseServerImpl(RepoAngajat iRepositoryEmployee, RepoCursa iRepositoryCursa, RepoInscriere iRepositoryInscriere, RepoPilot iRepositoryPilot) {
        this.iRepositoryEmployee = iRepositoryEmployee;
        this.iRepositoryCursa = iRepositoryCursa;
        this.iRepositoryInscriere = iRepositoryInscriere;
        this.iRepositoryPilot = iRepositoryPilot;
        loggedEmployees = new ConcurrentHashMap<>();
    }

    @Override
    public synchronized Angajat login(Angajat employee, ICursaObserver client) throws ProjectException {
        Angajat employee1 = iRepositoryEmployee.findByUsername(employee.getUsername());
        System.out.println(employee1);
        if(employee1 != null){
            if(loggedEmployees.get(employee.getUsername()) != null) {
                throw new ProjectException("Employee already logged in.");
            }
            if(!employee1.getPassword().equals(employee.getPassword())){
                throw new ProjectException("Wrong password.");
            }
            loggedEmployees.put(employee1.getUsername(), client);
            return employee1;
        }else{
            throw new ProjectException("Authentication failed.");
        }
    }

    @Override
    public synchronized void logout(Angajat employee, ICursaObserver client) throws ProjectException {
        if(loggedEmployees.get(employee.getUsername()) != null){
            loggedEmployees.remove(employee.getUsername());
        }
    }

    @Override
    public synchronized Iterable<Cursa> getCurse() throws ProjectException {
        return (List<Cursa>) iRepositoryCursa.getAll();
    }
    @Override
    public synchronized Iterable<Pilot> getPiloti() throws ProjectException
    {
        return (List<Pilot>) iRepositoryPilot.getAll();
    }


    public List<Pilot> getInscrisiEchipa(String nume, int capacitate)
    {
        Iterable<Inscriere> insc= iRepositoryInscriere.getAll();
        Iterable<Pilot> pil= iRepositoryPilot.getAll();
        Cursa cursa=iRepositoryCursa.findByCapacitate(capacitate);
        List<Inscriere> rez1 = new ArrayList<>();
        List<Pilot> rez2=new ArrayList<>();
        insc.forEach(x->{
            if(x.Cursa==cursa.getId()) {
                rez1.add(x);
            }
        });
        rez1.forEach(x->{
            pil.forEach(y->
            {
                if(x.getPilot()==y.Id && y.getNumeEchipa().equals(nume))
                    rez2.add(y);
            });
        });
        return rez2;
    }

    public Inscriere addInscriere(int id,String nume, String numeEchipa, int capacitate)
    {
        Cursa c=iRepositoryCursa.findByCapacitate(capacitate);
        Inscriere inscriere=new Inscriere(0,0);
        long id1=id;
        if(iRepositoryPilot.findOne(id1)==null)
        {
            iRepositoryPilot.add(new Pilot(id1,nume,numeEchipa));
        }
        Iterable<Pilot> pil=iRepositoryPilot.getAll();
        pil.forEach(x->
        {
            if(x.getNumeEchipa().equals(numeEchipa))
            iRepositoryInscriere.add(new Inscriere(x.getId(),c.getId()));

        });
        inscriere.setPilot(id1);
        inscriere.setCursa(c.getId());
        if(inscriere.getPilot()==0) {
            return null;
        }
        notifyClients(iRepositoryPilot.findOne(id1));
        return inscriere;
    }
    private void notifyClients(Pilot pilot){
        ExecutorService executor = Executors.newFixedThreadPool(defaultThreadsNo);
        for(String clientLogged : loggedEmployees.keySet()){
            ICursaObserver client = loggedEmployees.get(clientLogged);
            if(client != null) {
                executor.execute(() -> {
                    try {
                        System.out.println("Notifying [" + clientLogged + "]");
                        synchronized (client) {
                            client.inscriereAdded(pilot);
                        }
                        System.out.println("Notified");
                    } catch (ProjectException e) {
                        System.err.println("Error notifying client " + e);
                    }
                });
            }
        }
        executor.shutdown();
    }
}

File: CurseCS/Server/CurseServerImpl.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using curse.Model;
using curse.Services;
using System;
using curse.Persistance;

namespace curse.Server
{
    public class CurseServerImpl: ICurseServices
    {
        public IRepoAngajat userRepository;
        public IRepoInscriere inscriereRepository;
        public IRepoPilot pilotRepository;
        private readonly IDictionary <String, ICurseObserver> loggedClients;



    public CurseServerImpl(IRepoAngajat repo, IRepoInscriere messageRepo, IRepoPilot pilotRepository) {
        this.userRepository = repo;
        this.inscriereRepository = messageRepo;
        this.loggedClients=new Dictionary<String, ICurseObserver>();
        this.pilotRepository = pilotRepository;
    }

    public  void login(Angajat user, ICurseObserver client)  {
        Angajat userOk = userRepository.FindByUsername(user.Username);
            
        if (userOk!=null){
            if(loggedClients.ContainsKey(user.Username))
                throw new CurseException("User already logged in.");
            loggedClients[user.Username]= client;
            
        }else
            throw new CurseException("Authentication failed.");
    }


       public void notifyClients(String username, Pilot pilot)
       {
           ICurseObserver client = loggedClients[username];
           client.updateInscrieri(pilot,username);
           foreach (String us in loggedClients.Keys)
           {
               if (!us.Equals(username))
               {
                    ICurseObserver concursClient = loggedClients[us];
                    Task.Run(() => concursClient.updateInscrieri(pilot,us));
               }
           }
       }

        public  void logout(Angajat user, ICurseObserver client) 
        {
        ICurseObserver localClient=loggedClients[user.Username];
        if (localClient==null)
            throw new CurseException("User "+user.Id+" is not logged in.");
        loggedClients.Remove(user.Username);
        
        }

        public void AddInscriere(Inscriere insc,string username)
        {
            Console.WriteLine("INTRAT IN SERVER");
            try
            {
                if (loggedClients.ContainsKey(username))
                {
                    
                    inscriereRepository.Add(insc);
                    long id = insc.getPilot();
                    Pilot pilot = pilotRepository.FindOne(id);
                    notifyClients(username, pilot);
                    Console.WriteLine("afiseaza mesaj fara exceptie");

                }
                else
                    throw new CurseException("User not logged in!");
            }
            catch(Exception e)
            {
                throw new CurseException(e.Message);
            }
        }

        public Cursa[] GetCurse()
        {
            RepoCursa repo = new RepoCursa();
            IEnumerable<Cursa> curse = repo.GetAll();
            List<Cursa> rez=new List<Cursa>();
            foreach(var el in curse)
            {
                rez.Add(el);
            }
            return rez.ToArray();
        }
        public Pilot[] GetPiloti()
        {
            RepoPilot repo = new RepoPilot();
            IEnumerable<Pilot> piloti = repo.GetAll();
            List<Pilot> rez = new List<Pilot>();
            foreach(var el in piloti)
            {
                rez.Add(el);
            }
            return rez.ToArray();
        }
        public Pilot[] GetPilotiInscrisi(string nume, string numeEchipa, int capacitate)
        {
            RepoInscriere repoInscriere = new RepoInscriere();
            RepoPilot repoPilot = new RepoPilot();
            RepoCursa repoCursa = new RepoCursa();
            IEnumerable<Inscriere> insc = repoInscriere.GetAll();
            IEnumerable<Pilot> pil = repoPilot.GetAll();
            Cursa curse = repoCursa.FindByCapacitate(capacitate);

            List<Inscriere> rez1 = new List<Inscriere>();
            List<Pilot> rez2 = new List<Pilot>();
            if (curse != null)
            {
                foreach (var el in insc)
                {
                    if (curse.Id == el.Cursa)
                        rez1.Add(el);


                }
                foreach (var c in rez1)
                {
                    foreach (var el in pil)
                    {
                        if (c.Pilot == el.Id && el.numeEchipa == numeEchipa)
                            rez2.Add(el);
                    }
                }
            }
            return rez2.ToArray();
        }
    }
}
